### PR TITLE
Access token Publisher is too specific

### PR DESCRIPTION
**Pull request by junnae**
related to https://github.com/techcloud0/bankaxept-epayment-platform/issues/1357
Support for grantType, scope
apimKey optional
Created a builder to make it easier

### CODE REVIEW

File: java/access-token-client/src/main/java/no/bankaxept/epayment/client/accesstoken/AccessTokenRetriever.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package no.bankaxept.epayment.client.accesstoken;
+
+import no.bankaxept.epayment.client.base.accesstoken.AccessTokenPublisher;
+import no.bankaxept.epayment.client.base.accesstoken.AccessTokenSubscriber;
+import no.bankaxept.epayment.client.base.accesstoken.ScheduledAccessTokenPublisher;
+import no.bankaxept.epayment.client.webflux.WebFluxClient;
+
+import java.time.Duration;
+
+public class AccessTokenRetriever {
```
Comment (junnae) [inline java/access-token-client/src/main/java/no/bankaxept/epayment/client/accesstoken/AccessTokenRetriever.java:10]: I found it easier adding this token retriever like a module in this repo, so I could add the webflux-client with it

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java [hunk 2]
```diff
@@ -27,21 +36,41 @@ public Instant getExpiry() {
...
 
-        static AccessToken parse(String input) {
+        static AccessToken parse(String input, Clock clock) {
             var tokenMatcher = tokenPattern.matcher(input);
             var expiryMatcher = expiryPattern.matcher(input);
-            if (!tokenMatcher.find() || !expiryMatcher.find()) {
-                throw new IllegalArgumentException("Could not parse token or expiry"); //onError
+            var expirySecondsFromStartMatcher = expirySecondsFromStartPattern.matcher(input);
+            if ( !tokenMatcher.find()) {
+                throw new IllegalArgumentException("Could not parse token: " + input);
```
Comment (junnae) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java:65]: For debugging it was nice to output the whole token.. But i guess this could be bad to log, someone could technically use it?

Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java]: I removed input from the exception agai

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 3]
```diff
@@ -29,19 +29,38 @@ public class ScheduledAccessTokenPublisher implements AccessTokenPublisher, Flow
 
     private final Queue<Flow.Subscriber<? super String>> subscribers = new LinkedBlockingQueue<>();
 
+    private String grantType;
+    private String scopes;
```
Comment (junnae) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:33]: So Thomas suggested these shouldn't just be strings, or set to any value by the user.. But I spent so long on this, I didn't really do it properly yet. These should probably be enums you give a list of..?

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 4]
```diff
@@ -29,19 +29,38 @@ public class ScheduledAccessTokenPublisher implements AccessTokenPublisher, Flow
...
-        this.headers = createHeaders(apimKey, username, password);
+        this.headers = createHeaders(apimKey, username, password, false);
         this.clock = clock;
         this.scheduler = scheduler;
         this.httpClient = httpClient;
         scheduleFetch(0);
     }
 
-    private static LinkedHashMap<String, List<String>> createHeaders(String apimKey, String username, String password) {
+    public ScheduledAccessTokenPublisher(String uri, String id, String secret, String scopes, String grantType, HttpClient httpClient) {
```
Comment (junnae) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:44]: Maybe it's a bit messy to have two different constructors with different fields like this.. Would it be better to split into different classes or something else?

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java [hunk 5]
```diff
@@ -57,4 +57,11 @@ public boolean equals(Object o) {
     public int hashCode() {
         return Objects.hash(status);
     }
+
+    @Override
+    public String toString() {
```
Comment (junnae) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java:62]: We had an issue where it was hard to debug stuff because we can't see what http error status was returned. This solves this, but also exposes a bit more information than we wanted by abstracting the http errors in the first place..

File: java/base-client/src/test/resources/token-response.json [hunk 6]
```diff
@@ -1,4 +1,4 @@
 {
-  "expiresOn": 1645437609741,
-  "accessToken": "a-token"
-}
\ No newline at end of file
+  "expires_on": 1673262714,
```
Comment (junnae) [inline java/base-client/src/test/resources/token-response.json:2]: we actually both had the wrong format on these (should be expiry_on not expiryOn) AND it was epoch seconds and NOT MILLIseconds, which I assumed 🤔

File: java/base-client/src/test/resources/token-response2.json [hunk 7]
```diff
@@ -1,5 +1,5 @@
 {
-  "someOtherField": "abc",
-  "expiresOn": 123,
-  "accessToken": "a-token"
-}
\ No newline at end of file
+  "some_other_field": "abc",
+  "expires_on": 123, "some_field": "bca",
```
Comment (junnae) [inline java/base-client/src/test/resources/token-response2.json:3]: I'm not sure how/if we should even try to handle if expiry_in and expiry_on are somehow contradicting each other. I assume not

File: java/client-test/src/test/java/no/bankaxept/epayment/test/client/base/AccessTokenPublisherTest.java [hunk 8]
```diff
@@ -56,7 +59,7 @@ public void should_schedule_on_startup_then_new_on_success() throws IOException
         accessTokenProcessor = new ScheduledAccessTokenPublisher("uri", "key", "username", "password", clock, schedulerMock, httpClientMock);
         accessTokenProcessor.subscribe(subscriberMock);
         verify(schedulerMock).schedule(any(Runnable.class), eq(0L), eq(TimeUnit.MILLISECONDS));
-        verify(schedulerMock, Mockito.after(2000)).schedule(any(Runnable.class), eq(599999L), eq(TimeUnit.MILLISECONDS));
+        verify(schedulerMock, Mockito.after(2000)).schedule(any(Runnable.class), and(gt(1189000L), lt(1190000L)) , eq(TimeUnit.MILLISECONDS));
```
Comment (junnae) [inline java/client-test/src/test/java/no/bankaxept/epayment/test/client/base/AccessTokenPublisherTest.java:62]: was a pain to get "rougly this time".. But this *works* at least. Not sure if my precision is too loose

**Review by thxmasj — CHANGES_REQUESTED** [review]
The purpose of this development kit is to support EPP integrators. This PR adds code which purpose is to support EPP internals only, which defeats that purpose.

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java [hunk 9]
```diff
@@ -27,21 +36,41 @@ public Instant getExpiry() {
...
+            }
+            if (expirySecondsFromStartMatcher.find()) {
```
[the remark this answers is not in the document]
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java:59]: What is the deal with this change, anyway? Is our access-token-api buggy and not following the OAuth2 standard by using invalid parameters (yes)? That deserves a separate problem description (issue) and fix (PR).

[the remark this answers is not in the document]
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java:59]: [`expires_in` is the OAuth2 standard](https://www.rfc-editor.org/rfc/rfc6749#section-5.1)

File: java/access-token-client/src/main/java/no/bankaxept/epayment/client/accesstoken/AccessTokenRetriever.java (hunk 1 above)
Comment (junnae) [reply java/access-token-client/src/main/java/no/bankaxept/epayment/client/accesstoken/AccessTokenRetriever.java]: I have moved it back to platform

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java:59]: https://github.com/techcloud0/bankaxept-epayment-development-kit/blob/42d81c7d9a833a34291c6bbc4e61ea82773f47f8/openapi/access-token/server.yaml#L56 🤷  I looked at our yaml

**Comment by junnae** [comment]
I have removed the access-token-client, so there shouldn't be any direct epp code now. It's just some bug fixes/facilitating, and making the token retrieving configurable

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java:59]: We obviously need to fix the access-token-api. Did you look at that separately?

File: java/base-client/src/test/resources/token-response2.json (hunk 7 above)
Comment (thxmasj) [reply java/base-client/src/test/resources/token-response2.json:3]: See other comment. There's no `expiry_on` in the OAuth2 spec.

**Comment by thxmasj** [comment]
> I have removed the access-token-client, so there shouldn't be any direct epp code now. It's just some bug fixes/facilitating, and making the token retrieving configurable

Can't see you have removed it.

**Comment by junnae** [comment]
Right, you have to click "delete" twice in intellij. It's gone now

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java (hunk 5 above)
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/http/HttpStatus.java:62]: https://github.com/orgs/techcloud0/projects/2/views/2?filterQuery=status&pane=issue&itemId=10447149

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java [hunk 10]
```diff
@@ -36,9 +35,8 @@ public Instant getExpiry() {
         return expiry;
     }
 
-    public long secondsUntilTenMinutesBeforeExpiry(Clock clock) {
-        Instant tenMinutesBeforeExpiry = expiry.minus(10, ChronoUnit.MINUTES);
-        return Duration.between(clock.instant(), tenMinutesBeforeExpiry).toSeconds();
+    public long tenSecondsBeforeExpiry(Clock clock) {
```
Comment (junnae) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java:38]: Does this need to be configurable? I couldn't find any definitive standard for this value

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 11]
```diff
@@ -52,7 +49,7 @@ private void scheduleFetch(long seconds) {
 
     private void fetchNewToken() {
         if (shutDown) return;
-        httpClient.post(uri, new SinglePublisher<>("", fetchExecutor), headers).subscribe(this);
+        httpClient.post(uri, new SinglePublisher<>(body, fetchExecutor), headers).subscribe(this);
```
Comment (thxmasj) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:52]: Ok, makes me think we never used this code...

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 12]
```diff
@@ -63,7 +60,7 @@ public void onSubscribe(Flow.Subscription subscription) {
     @Override
     public void onNext(HttpResponse item) {
         if (!item.getStatus().is2xxOk()) {
-            onError(new HttpStatusException(item.getStatus(), "Error when fetching token"));
+            onError(new HttpStatusException(item.getStatus(), "Error when fetching token:"));
```
Comment (thxmasj) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:63]: Why this change?

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java (hunk 10 above)
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java:38]: We can add configurability later, if we see need for it. Could you keep the `secondsUntil` prefix still? Otherwise an Instant should be returned...

**Review by thxmasj — CHANGES_REQUESTED** [review]

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java (hunk 10 above)
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java]: Sure, changed. `secondsUntilTenSecondsBeforeExpiry` is not the the most elegant name though

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 13]
```diff
@@ -108,12 +105,82 @@ public void shutDown() {
...
             synchronized (subscribers) {
                 subscribers.add(subscriber);
             }
         }
     }
+
+    public static class Builder {
+        private String uri;
+        private HttpClient httpClient;
+        private ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
```
[the remark this answers is not in the document]
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:127]: We do use it, as it was almost always set to this value

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 14]
```diff
@@ -108,12 +105,82 @@ public void shutDown() {
...
+            headers.put("Authorization", List.of("Basic " + Base64.getEncoder().encodeToString((id + ":" + secret).getBytes(StandardCharsets.UTF_8))));
+            return this;
+        }
+
+        public Builder grantType(GrantType grantType) {
+            prefix().append("grant_type=").append(grantType);
+            return this;
+        }
+
+        public Builder scopes(List<Scope> scopes) {
```
[the remark this answers is not in the document]
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:159]: `OAuth does not define any particular values for scopes, since it is highly dependent on the service's internal architecture and needs.`.
That's why I did it like this, and implemented BankIDScope in platform project

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 15]
```diff
@@ -108,12 +105,82 @@ public void shutDown() {
...
+        }
+
+        private StringBuilder prefix(){
+            return body.length() > 0 ? body.append("&") : body;
+        }
+
+        public ScheduledAccessTokenPublisher build() {
+            if(body.length() > 0) {
+                headers.put("Content-type", List.of("application/x-www-form-urlencoded"));
+            }
```
[the remark this answers is not in the document]
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java]: Now it throws `IllegalArgumentException` if you don't set grantType. Is that fine?

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/Scope.java [hunk 16]
```diff
@@ -0,0 +1,6 @@
+package no.bankaxept.epayment.client.base.accesstoken;
+
+
+public interface Scope {
+    String getValue();
+}
```
[the remark this answers is not in the document]
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/Scope.java:6]: But why? https://github.com/techcloud0/bankaxept-epayment-platform/blob/4fe98a35414476139d7618a272679786b0ae1609/lib/bankid/src/main/java/no/bankaxept/ebax/bankid/BankIDScope.java
This is how I use it. How else would we do it?

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 17]
```diff
@@ -108,12 +105,82 @@ public void shutDown() {
...
-        if (token != null && token.getExpiry().isBefore(clock.instant()))
+        if (token != null){
+            if(token.getExpiry().isBefore(clock.instant())) {
+                var exception = new IllegalStateException("Token already expired at: " + token.getExpiry());
+                subscriber.onError(exception);
+                onError(exception);
+
+            }
             subscriber.onNext(token.getToken());
+        }
```
[the remark this answers is not in the document]
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java]: You're right. This should be reverted now. This was a leftover from debugging/problems I had.

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java (hunk 12 above)
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java]: I reverted this also. I had expanded this to log more (like response http status, but this `:` was a leftover)

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java (hunk 11 above)
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:53]: Yeah, we never had a body before 🤷

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:121]: The builder still allows a custom one, so we should not create this one in that case. Alternatively remove support for custom ones.

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:153]: Why not just use a String in this generic code, then? The interface is just a supplier now. We can still use enums on the consumers.

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/Scope.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/Scope.java:6]: You could just as well use BankIdScope::getValue, which is a String, and use that for the builder. Type safety is not affected.

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/BaseClient.java [hunk 18]
```diff
@@ -116,13 +116,25 @@ public Builder withSuppliedToken(Supplier<String> tokenSupplier) {
...
+        public Builder withScheduledToken(String id, String secret, Clock clock) {
+            this.tokenPublisher = new ScheduledAccessTokenPublisher.Builder()
+                    .httpClient(httpClient)
+                    .uri("/bankaxept-epayment/access-token-api/v1/accesstoken")
+                    .grantType(GrantType.client_credentials)
+                    .credentials(id, secret)
+                    .apimKey(apimKey)
+                    .clock(clock)
+                    .build();
             return this;
```
Comment (thxmasj) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/BaseClient.java:138]: The one sets grant type, the other not, which is inconsistent. Also, as client_credentials is the only one we support, we don't need to include it in the builder. You could perhaps set it implicitly with the `credentials(String, String)`?

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java]: 👍

**Review by thxmasj — CHANGES_REQUESTED** [review]

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 19]
```diff
@@ -108,12 +105,82 @@ public void shutDown() {
...
+            return this;
+        }
+
+        public Builder credentials(String id, String secret) {
+            headers.put("Authorization", List.of("Basic " + Base64.getEncoder().encodeToString((id + ":" + secret).getBytes(StandardCharsets.UTF_8))));
+            return this;
+        }
+
+        public Builder grantType(GrantType grantType) {
+            prefix().append("grant_type=").append(grantType);
```
[the remark this answers is not in the document]
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java]: 👍

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:121]: Right, the problem is not with the _style_, but the fact that it gets created anyway? I moved creation of a default value into the constructor

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:153]: Yeah, makes sense. Changed it to string

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/Scope.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/Scope.java:6]: It is deleted

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/BaseClient.java (hunk 18 above)
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/BaseClient.java:138]: I set it implicitly, renamed it to `clientCredentials`, and made `grantType()` private. Does that make sense?

Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/BaseClient.java]: Or do we still want to set it manually?

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 20]
```diff
@@ -23,27 +25,23 @@ public class ScheduledAccessTokenPublisher implements AccessTokenPublisher, Flow
...
-    private final LinkedHashMap<String, List<String>> headers;
+    private final Map<String, List<String>> headers;
+    private final String body;
 
     private final AtomicReference<AccessToken> atomicToken = new AtomicReference<>();
 
     private final Queue<Flow.Subscriber<? super String>> subscribers = new LinkedBlockingQueue<>();
 
-    public ScheduledAccessTokenPublisher(String uri, String apimKey, String username, String password, Clock clock, ScheduledExecutorService scheduler, HttpClient httpClient) {
+    private ScheduledAccessTokenPublisher(String uri, Map<String, List<String>> headers, String body, Clock clock, ScheduledExecutorService scheduler, HttpClient httpClient) {
```
Comment (junnae) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:35]: I changed this to a generic Map, I don't see why it needed to be a specific implementation? 🤔

Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:35]: order of headers shouldn't matter

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 21]
```diff
@@ -116,4 +114,74 @@ public void subscribe(Flow.Subscriber<? super String> subscriber) {
...
         }
     }
+
+    public static class Builder {
+        private String uri;
+        private HttpClient httpClient;
+        private ScheduledExecutorService scheduler;
+        private Clock clock = Clock.systemDefaultZone();
+
+        private final Map<String, List<String>> headers = new HashMap<>(Map.of("Content-type", List.of("application/x-www-form-urlencoded")));
```
Comment (junnae) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:124]: I put this value in here now, since it should always be set now. also avoids inserting it multiple times if you build() again, though it shouldn't matter too much

**Review by thxmasj — APPROVED** [review]

File: java/base-client/src/test/resources/token-response.json [hunk 22]
```diff
@@ -1,4 +1 @@
-{
-  "expires_in": 3600,
-  "access_token": "a-token"
-}
\ No newline at end of file
+{"access_token":"a-token","expires_in":3600,"refresh_expires_in":0,"token_type":"bearer","id_token":"id_token","not_before_policy":"","scope":""}
```
Comment (junnae) [inline java/base-client/src/test/resources/token-response.json:1]: this is copied from response in dev

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java [hunk 23]
```diff
@@ -108,12 +106,82 @@ public void shutDown() {
     @Override
     public void subscribe(Flow.Subscriber<? super String> subscriber) {
         var token = atomicToken.get();
-        if (token != null && token.getExpiry().isBefore(clock.instant()))
+        if (token != null && token.getExpiry().isAfter(clock.instant()))
```
Comment (junnae) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:109]: don't know how this happened

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java [hunk 24]
```diff
@@ -36,13 +35,12 @@ public Instant getExpiry() {
...
-    public long secondsUntilTenMinutesBeforeExpiry(Clock clock) {
-        Instant tenMinutesBeforeExpiry = expiry.minus(10, ChronoUnit.MINUTES);
-        return Duration.between(clock.instant(), tenMinutesBeforeExpiry).toSeconds();
+    public long secondsUntilTenSecondsBeforeExpiry(Clock clock) {
+        return Duration.between(clock.instant(), expiry.minusSeconds(10)).toSeconds();
     }
 
     private static class Parser {
-        private static final Pattern tokenPattern = Pattern.compile("\"access_token\"\\s*:\\s*\"(.*)\"");
+        private static final Pattern tokenPattern = Pattern.compile("\"access_token\"\\s*:\\s*\"(.*?)\"");
```
Comment (thxmasj) [inline java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java:43]: This needs explanation (as comment in code?). Why should the pattern accept empty values?

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java (hunk 23 above)
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:109]: Do we lack test coverage, perhaps?

**Review by thxmasj — CHANGES_REQUESTED** [review]

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java (hunk 23 above)
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:109]: So the thing is.. We *DO* have a test.. But I think that the problem is that when we fetch a token, it's pushed to all subscribers without verifying the date, but when we subscribe to an already "cached" token, we do this check..
I changed the test to check twice, this will fail with the old code

Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/ScheduledAccessTokenPublisher.java:109]: [/AccessTokenPublisherTest.java](https://github.com/techcloud0/bankaxept-epayment-development-kit/pull/57#diff-164fcdf7ebfb787858fc55714aa8996dc5b171c8947da3e8c4a79f240ad5a805)

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java (hunk 24 above)
Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java]: I don't think `.+?` makes sense. `.+` is good if you want to make sure the string is not empty.

Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java]: the problem isn't making sure the string isn't empty, it's when the whole response is on the same line

File: java/client-test/src/test/java/no/bankaxept/epayment/test/client/base/AccessTokenPublisherTest.java [hunk 25]
```diff
@@ -52,30 +55,44 @@ public void tearDown(){
...
         doReturn(new SinglePublisher<>(new HttpResponse(200, tokenResponse()), executor)).when(httpClientMock).post(eq("uri"), any(), any());
-        accessTokenProcessor = new ScheduledAccessTokenPublisher("uri", "key", "username", "password", clock, schedulerMock, httpClientMock);
+        accessTokenProcessor = createPublisher();
         accessTokenProcessor.subscribe(subscriberMock);
         verify(schedulerMock).schedule(any(Runnable.class), eq(0L), eq(TimeUnit.SECONDS));
-        verify(schedulerMock, Mockito.after(2000)).schedule(any(Runnable.class), eq(3000L), eq(TimeUnit.SECONDS));
+        verify(schedulerMock, Mockito.after(2000)).schedule(any(Runnable.class), eq(3590L), eq(TimeUnit.SECONDS));
         verify(subscriberMock).onNext("a-token");
+        accessTokenProcessor.subscribe(anotherSubscriberMock);
+        verify(anotherSubscriberMock).onNext("a-token");
```
Comment (thxmasj) [inline java/client-test/src/test/java/no/bankaxept/epayment/test/client/base/AccessTokenPublisherTest.java:64]: Can you comment what's going on here so other developers might get it, too?

File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java (hunk 24 above)
Comment (junnae) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java]: I have added a test that fails if this is not lazy

Comment (thxmasj) [reply java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java]: Sorry, my bad. `.+?` makes sense, as it removes greediness from `.+`. 👍

File: java/client-test/src/test/java/no/bankaxept/epayment/test/client/base/AccessTokenPublisherTest.java (hunk 25 above)
Comment (junnae) [reply java/client-test/src/test/java/no/bankaxept/epayment/test/client/base/AccessTokenPublisherTest.java]: I added a comment here

**Review by thxmasj — APPROVED** [review]
